pgx/src/lib.rs

```diff
@@ -307,6 +308,38 @@ macro_rules! pg_sql_graph_magic {
     };
 }
 
+pub(crate) static UTF8DATABASE: Lazy<Utf8Compat> = Lazy::new(|| {
+    let encoding_int = unsafe { pgx_pg_sys::GetDatabaseEncoding() };
+    match encoding_int as core::ffi::c_uint {
+        pg_sys::pg_enc_PG_UTF8 => Utf8Compat::Yes,
+        // The 0 encoding. It... may be UTF-8
+        pg_sys::pg_enc_PG_SQL_ASCII => Utf8Compat::Maybe,
```

Comment:
This (and the logic that plays off it) optimistically assumes that PGX _may_ be active in a UTF-8-conformant database, and that if this is false, normal UTF-8 validation will sniff this out pretty quickly. This implies a performance penalty, however.